### PR TITLE
feat(binder): derive expr's output column names

### LINKED ISSUE #3812: frontend: Derive output column names
In PG
```sql
postgres=# select abs(1);
 abs
-----
   1
(1 row)

postgres=# select count(1);
 count
-------
     1
(1 row)
```

In risingwave
```sql
dev=> select abs(1);
 ?column?
----------
        1
(1 row)

dev=> select count(1);
 ?column?
----------
        1
(1 row)
```

related but seems different: https://github.com/singularity-data/risingwave/issues/3690

### PR DESCRIPTION
**Pull request by st1page**
8I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
 
fix #3812

Also remove some non-pg types in sqlparser

### Types of user-facing changes

* SQL commands, functions, and operators

### Release note

Improve the derivation of column name derivation to be compatible with PostgreSQL.

By the way, do we need to explain the derivation of materialized views' column names somewhere in our documents?  

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Review by xxchan — APPROVED** [review]

**Comment by xxchan** [comment]
@st1page I did a refactor

**Comment by st1page** [comment]
> @st1page I did a refactor

Good work, thanks!

**Comment by xxchan** [comment]
fixing failure...

File: src/frontend/src/handler/create_mv.rs [hunk 1]
```diff
@@ -150,6 +154,20 @@ It only indicates the physical clustering of the data, which may improve the per
     }
 }
 
+fn ordinal(i: usize) -> String {
```
Comment (fuyufjh) [inline src/frontend/src/handler/create_mv.rs:157]: So considerate!  🥰

File: src/frontend/src/handler/create_mv.rs [hunk 2]
```diff
@@ -214,36 +232,40 @@ pub mod tests {
...
+        let sql = "create materialized view mv0 as select count(x) from t";
+        frontend.run_sql(sql).await.unwrap();
+
+        // Same aggregations without alias is forbidden, because it make the same column name.
+        let sql = "create materialized view mv1 as select count(x), count(*) from t";
         let err = frontend.run_sql(sql).await.unwrap_err();
         assert_eq!(
             err.to_string(),
-            "Bind error: An alias must be specified for an expression"
+            "Invalid input syntax: column \"count\" specified more than once"
```
Comment (fuyufjh) [inline src/frontend/src/handler/create_mv.rs:252]: This test case seems to be a good example for documents.

**Review by fuyufjh — APPROVED** [review]
LGTM

File: src/frontend/src/binder/select.rs [hunk 3]
```diff
@@ -180,21 +180,9 @@ impl Binder {
...
-                        ),
-                        Expr::FieldIdentifier(field_expr, idents) => (
-                            self.bind_single_field_column(*field_expr.clone(), &idents)?,
-                            idents.last().map(|ident| ident.real_value()),
-                        ),
-                        _ => (self.bind_expr(expr)?, None),
-                    };
-                    select_list.push(select_expr);
+                    let bound = self.bind_expr(expr.clone())?;
+                    let alias = derive_alias(&expr);
```
Comment (xiangjinwu) [inline src/frontend/src/binder/select.rs:184]: nit: clone can be avoided by swapping these 2 statements.

File: src/frontend/src/binder/select.rs [hunk 4]
```diff
@@ -389,3 +377,51 @@ impl Binder {
...
+        AstDataType::Float(Some(0 | 54..)) => unreachable!(),
+        AstDataType::SmallInt => "int2".to_string(),
+        AstDataType::Int => "int4".to_string(),
+        AstDataType::BigInt => "int8".to_string(),
+        AstDataType::Boolean => "bool".to_string(),
+        AstDataType::Date => "date".to_string(),
+        AstDataType::Time(tz) => format!("time{}", if *tz { "z" } else { "" }),
+        AstDataType::Timestamp(tz) => {
+            format!("timestamp{}", if *tz { "z" } else { "" })
+        }
```
Comment (xiangjinwu) [inline src/frontend/src/binder/select.rs:413]: The suffix is `tz` rather than `z`

**Review by xiangjinwu — COMMENTED** [review]
rest lgtm

File: src/frontend/src/binder/select.rs [hunk 5]
```diff
@@ -389,3 +377,51 @@ impl Binder {
...
 }
+
+fn derive_alias(expr: &Expr) -> Option<String> {
+    match expr.clone() {
+        Expr::Identifier(ident) => Some(ident.real_value()),
+        Expr::CompoundIdentifier(idents) => idents.last().map(|ident| ident.real_value()),
+        Expr::FieldIdentifier(_, idents) => idents.last().map(|ident| ident.real_value()),
+        Expr::Function(func) => Some(func.name.real_value()),
+        Expr::Case { .. } => Some("case".to_string()),
+        Expr::Cast { expr, data_type } => derive_alias(&expr).or(data_type_to_alias(&data_type)),
```
Comment (BugenZhao) [inline src/frontend/src/binder/select.rs:388]: No warning here? 🤔

```suggestion
        Expr::Cast { expr, data_type } => derive_alias(&expr).or_else(|| data_type_to_alias(&data_type)),
```

**Comment by st1page** [comment]
> By the way, do we need to explain the derivation of materialized views' column names somewhere in our documents?

1. for columns the user needed, the column name of the result follows the same logic as PostgreSQL. Unfortunately, I do not find the PostgreSQL document about that. I think it is too verbose to give a detailed doc.
2. for hidden columns, I think we can not give any promise for the column names.

**Pull request merged**